caffe2/CMakeLists.txt

```diff
@@ -979,47 +1009,73 @@ if (MSVC AND NOT BUILD_SHARED_LIBS)
   #
   # NB: This must be target_compile_definitions, not target_compile_options,
   # as the latter is not respected by nvcc
-  target_compile_definitions(torch PUBLIC "AT_CORE_STATIC_WINDOWS=1")
+  target_compile_definitions(torch_cpu PUBLIC "AT_CORE_STATIC_WINDOWS=1")
 endif()
 if (MSVC AND BUILD_SHARED_LIBS)
   # ONNX is linked statically and needs to be exported from this library
   # to be used externally. Make sure that references match the export.
-  target_compile_options(torch PRIVATE "-DONNX_BUILD_MAIN_LIB")
+  target_compile_options(torch_cpu PRIVATE "-DONNX_BUILD_MAIN_LIB")
 endif()
 
-install(TARGETS torch EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+caffe2_interface_library(torch_cpu torch_cpu_library)
 
+if (USE_CUDA)
+  caffe2_interface_library(torch_cuda torch_cuda_library)
+elseif (USE_ROCM)
+  caffe2_interface_library(torch_hip torch_hip_library)
+endif()
 
 caffe2_interface_library(torch torch_library)
+
+install(TARGETS torch_cpu torch_cpu_library EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
```

Comment:
Guess we need add these interface targets (torch_cpu / torch_cuda / etc) to TORCH_LIBRARIES variable in TorchConfig.cmake.in so that client code doesn't have to explicitly include them? (or I missed where it's happening already?)
We can fix it separately though (I think static library cmake is already kinda broken anyway)

Reply:
Nope, this is not necessary. This is because the `torch` target depends on `torch_cpu_library`, which means it gets the interface whole archive gobbeldygook. Or at least, I think that's what is happening: I'm mostly relying on CI to tell me if I got it right...

Reply:
I thought most of our CIs are building first-party targets with the main cmake build process. Do we have CI testing the 3rd-party code using the generated cmake in install/share/cmake/Torch? I'm trying to add one for mobile.

Reply:
Yes we do; this is what the custom operator tests test:

```
    # Build custom operator tests.
    CUSTOM_OP_BUILD="$PWD/../custom-op-build"
    CUSTOM_OP_TEST="$PWD/test/custom_operator"
    python --version
    SITE_PACKAGES="$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())')"
    mkdir "$CUSTOM_OP_BUILD"
--  pushd "$CUSTOM_OP_BUILD"
--  cmake "$CUSTOM_OP_TEST" -DCMAKE_PREFIX_PATH="$SITE_PACKAGES/torch" -DPYTHON_EXECUTABLE="$(which python)"
    make VERBOSE=1
--  popd
    assert_git_not_dirty
```

I know this because I broke it a few times working on this patchset :)

Reply:
Seems the test doesn't fully cover 3rd party build + static library case - at least it fails my test/mobile/op_deps project build.
I fixed it in a naive way and integrated that build test into mobile CI: #30769

Reply:
Thanks @ljk53 !